### PR TITLE
clickRoutine now in every widget and added CALL

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
Looking at the discussions in the issues, these features are pretty controversial.

I argued against them in the past because I thought users should know what happens when they click stuff, but I think we're already way past that. So I say: give game devs all the power and let them create amazing games!

It might be problematic to have them all in one PR but they are a bit interconnected.

Note that defining a `clickRoutine` in widgets that already have click behaviour disables the normal behaviour.

~~I added `propertyChangeRoutine`, `childAddRoutine` and `childRemoveRoutine` because those are exactly the events that the engine uses internally and so far are all I need.~~

~~- `propertyChangeRoutine` receives these variables: `property`, `oldValue` and `newValue`~~
~~- `childAddRoutine` receives variable `oldParentID` and a collection `child`~~
~~- `childRemoveRoutine` receives the collection `child`~~

_The new routines will be added in a separate PR._

For `CALL`, you can look at the button in the demo room. Defaults are: `{ widget: this.p('id'), routine: 'clickRoutine', 'return': true, arguments: {}, variable: 'result', collection: 'result' }`

- `return: false` ends the button execution after the `CALL`
- `arguments` can be used to inherit variables to the callee **in addition** to all currently set variables
- collections are inherited as well
- `routine` can be used to define custom properties like `myCustomClickRoutine` - they have to end on `Routine` though just to be easily recognizable (by humans and JSON editors)
- `variable` receives the value of the variable `result` in the callee so this can be used to receive data back from the callee
- `collection` is the same as `variable` but for a collection `result`
- ~~`iterations` defaults to 1 but can be used as a `for` loop - the callee will have a variable `iteration`~~

_`iterations` has been removed because we will add operations like `FOR` instead._

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3197/pr-test (or any other room on that server)

**Comment by shaunhillier** [comment]
Wow, now I want to rebuild Spendid :) can't wait to play with this.
The only suggestion I would have, is would it be possible to default the JSON for widgets to have their default behaviour visible. e.g. on a card have the click routine default to a flip? then nothing is hidden and you don't have to remember defaults.

**Comment by RaphaelAlvez** [comment]
I just want to point out that now a basicWidget is by far more powerful that a button. You can set a clickRoutine in the faces and have a button cycle through routines or even get random routines

**Comment by ArnoldSmith86** [comment]
> would it be possible to default the JSON for widgets to have their default behaviour visible

Maybe. :D

**Comment by shaunhillier** [comment]
I think that last commit has taken the test server down, I can't access pr-test or my own room.

**Comment by shaunhillier** [comment]
So I am having a play in http://212.47.248.129:3197/sh-test, I've noticed one thing.

I have added an "childAddRoutine" function to a holder and noticed that when I remove the second to last card it is triggered. I'm guessing that's because the pile is disbanded so the last card is technically added.

**Comment by robartsd** [comment]
I did have the idea that for visibility of functionality in edit mode, we could style widgets with non-default click behavior with a pseudo element that displays an icon representing the routine.

**Comment by robartsd** [comment]
> > would it be possible to default the JSON for widgets to have their default behaviour visible
> 
> Maybe. :D

I don't think copying the default behavior to every widget in the json would be a good idea, but a separate panel displaying the default properties being applied when in the json editor would be nice.

**Comment by larkob** [comment]
I suggest the following names for the new routines:
- changeRoutine
- enterRoutine
- leaveRoutine

For custom routines, we should allow for any name. In case they end with Routine, they could be recognized as such by the JSON editor, but I don't think this convenience feature should make it a requirement for the naming.

**Comment by RaphaelAlvez** [comment]
On top of that I recommend turning onLeave and onEnter into childLeaveRoutine and childEnterRoutine

**Comment by RaphaelAlvez** [comment]
We need to check what happens when you have an onEnter and a enterRoutine together

**Comment by larkob** [comment]
I've tested the `CALL` operation and found only a few issues:
* setting `variable` in the CALL operation to anything but "result" did not work - is this really just a default or a fixed variable name?
* setting `variables` had no noticeable effect, but I also don't quite understand what it does - can you provide an example of its proper use?

**Comment by larkob** [comment]
I just figured out the `variable` issue. It works as follows: The called routine always uses the variable name "result" in order to pass a value to its previous caller, but the calling routine can take this value via a differently named variable for subsequent operations.

File: client/js/widgets/button.js [hunk 1]
```diff
@@ -28,452 +24,6 @@ export class Button extends Widget {
...
-      }
-    }
-
-    if(this.p('debug'))
-      showOverlay('debugButtonOverlay');
-
-    batchEnd();
-  }
-
   evaluateInputOverlay(o, resolve, reject, go) {
```
Comment (UltimateGeek) [inline client/js/widgets/button.js:27]: Shouldn't the corresponding input overlay functionality also be moved to widget.js?

Comment (UltimateGeek) [reply client/js/widgets/button.js:27]: I added this to the room to test INPUT functionality using a `button`. It doesn't seem to work.

image

Once that works, could change the widget type from button to something else.

**Comment by larkob** [comment]
In our call we decided to remove `iterations` from CALL, and to rename `variables` to `arguments`.

**Comment by larkob** [comment]
Note from our call today: We discussed that the new clickRoutines would require significant changes in order to make everything work with async/await. We decided that we'd split this ticket into two parts. One to add a CALL button operation to the existing clickRoutine and another (at a later point) to add the new clickRoutines.

**Comment by robartsd** [comment]
> Note from our call today: We discussed that the new change/enter/leave routines would require significant changes in order to make everything work with async/await. We decided that we'd split this ticket into two parts. One to add a CALL button operation to the existing clickRoutine and another (at a later point) to add the change/enter/leave routines.

I'd be happy if this were split to 3 parts:
1) move clickRoutine code from button.js to widget.js
2) add CALL / support for named routines
3) add change/enter/leave routine triggers

I'd also be absolutely OK if we decided that change/enter/leave routines cannot ask for user input (the reason async/await is needed); but I recognize that this might make maintenance of routines code more difficult.

**Comment by ArnoldSmith86** [comment]
This should be done (I removed changeRoutine/enterRoutine/leaveRoutine for now) but the tests still fail. I assume the tests themselves need updating so you can probably already review the rest of the code and test the functionality.

I really want to get this merged ASAP because it's really a pain to keep up-to-date with button changes.

I will look at the tests myself if no-one else does (@UltimateGeek ? :) ) but for now I ran out of time.

**Comment by larkob** [comment]
Just to get people started on how CALL could be used, have a look at http://212.47.248.129:3197/larkob

**Review by RaphaelAlvez — DISMISSED** [review]
Maybe change the name of the collection "thisButon"

**Comment by ArnoldSmith86** [comment]
Which collection?

**Comment by RaphaelAlvez** [comment]
> Which collection?

It's a default collection containing only the item clicked. Check the checker piece on Lars room

**Comment by ArnoldSmith86** [comment]
So you mean renaming `thisButton` to something like `thisWidget`?

Sounds reasonable but that needs a version bump and I really don't want to wait much longer with this PR. The moved button code has high risk of breaking conflicts..

We can always rename it later though.

**Comment by RaphaelAlvez** [comment]
You're right

**Comment by robartsd** [comment]
JSON editor does not know about `clickRoutine` in `cardDefaults` (looks like it doesn't really know much about cardDefaults at all). If I manually add `clickRoutine` to `cardDefaults` `clickRoutine` commands for the JSON editor are displayed but attempting to use them crashes the client.

**Review by robartsd — APPROVED** [review]
I've tested Rummy Tiles (currently submitted version working around recent regression) and find no issues. I've also built a dice deck using a `clickRoutine` in `cardDefaults`. Works as expected.

**Comment by ArnoldSmith86** [comment]
> JSON editor does not know about `clickRoutine` in `cardDefaults` (looks like it doesn't really know much about cardDefaults at all). If I manually add `clickRoutine` to `cardDefaults` `clickRoutine` commands for the JSON editor are displayed but attempting to use them crashes the client.

I'll look at the crash. The rest is out of scope. Is there an issue for that?

**Comment by larkob** [comment]
So far, I couldn't find any problems in games. Some minor issues I discovered in JSON editor are:
* when you duplicate a widget in JSON editor, it says "ID xxxx is already in use" and then you have to refresh or go to some other widget first before you can continue editing the duplicated widget
* when you add a clickRoutine to a button (or any widget) it adds `"clickRoutine": null` instead of `"clickRoutine": []` and when you then try to add an operation like COMPUTE, the room crashes because it expects clickRoutine to be an array - this is probably the issue mentioned by robartsd

**Comment by rogerl50** [comment]
Will the debug property in a caller apply also to the callee? What will/should happen if the callee has `debug` set but the parent does not?

**Comment by robartsd** [comment]
> * when you add a clickRoutine to a button (or any widget) it adds `"clickRoutine": null` instead of `"clickRoutine": []` and when you then try to add an operation like COMPUTE, the room crashes because it expects clickRoutine to be an array - this is probably the issue mentioned by robartsd

I think it has additional issues if the clickRoutine is in cardDefaults. I ended up building my dice routine directly on a card, then moving it to cardDefaults.

**Comment by ArnoldSmith86** [comment]
Thanks for all the testing. I'll look into the results.

Did you test if the "ID already in use" also happens in production?

**Comment by larkob** [comment]
I did. The "ID already in use" issue does not happen in production.

**Comment by ArnoldSmith86** [comment]
> Will the debug property in a caller apply also to the callee? What will/should happen if the callee has `debug` set but the parent does not?

Currently it won't.

When this is merged, I will significantly improve `debug`.

**Comment by ArnoldSmith86** [comment]
Did I miss anything you reported?

**Review by larkob — APPROVED** [review]

**Comment by rogerl50** [comment]
Is there a way to return a collection?

**Comment by ArnoldSmith86** [comment]
No, never thought about that. Any suggestions about how that should work?

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -200,10 +241,543 @@ export class Widget extends StateManaged {
...
+            const inheritCollections = {};
+            for(const c in collections)
+              inheritCollections[c] = [ ...collections[c] ];
+            inheritCollections['caller'] = [ this ];
+            $('#debugButtonOutput').textContent += `\n\n\nCALLing: ${a.widget}.${a.routine}\n`;
+            variables[a.variable] = await widgets.get(a.widget).evaluateRoutine(a.routine, inheritVariables, inheritCollections, (depth || 0) + 1);
+          }
+        }
+        if(!a.return) {
+          $('#debugButtonOutput').textContent += '\n\n\nCALL without return. Ending evaluation.\n';
```
Comment (rogerl50) [inline client/js/widgets/widget.js:334]: Two questions:

1. What is the purpose of this line? It's evaluated in the context of the caller if I understand it correctly, so won't it always fire?
2. What is the purpose of the `return:false` property mentioned in the summary? I don't see `return` being used anywhere else in the code.

**Comment by rogerl50** [comment]
> No, never thought about that. Any suggestions about how that should work?

Since `result` is simply a distinguished variable name that always represents a returned variable, we could also define a distinguished collection name. 

I suppose this opens the door for other requests as well. For example, I could imagine an application where you would want to return a button reference, for example, but that fits into the existing variable model since `applyVariables` could get you to the actual button.

File: client/js/widgets/widget.js (hunk 2 above)
Comment (ArnoldSmith86) [reply client/js/widgets/widget.js:334]: The `result` is what the called routine can give back to its caller.

Setting `return` to `false` forces the calling routine to end after the `CALL` operation, turning it more into a GOTO or JUMP. And this place in the code is exactly where that happens: `if(!a.return) break;`

Comment (rogerl50) [reply client/js/widgets/widget.js:334]: I see. I didn't read the spec carefully enough. This makes sense.

**Comment by larkob** [comment]
How about the returned collection uses the same name as the variable, i.e. "result"? That would be easy to remember.

**Comment by larkob** [comment]
If anyone wants to have better idea of how exactly CALL works, I've documented the CALL operation on https://github.com/ArnoldSmith86/virtualtabletop/wiki/Upcoming-Changes

**Comment by rogerl50** [comment]
> How about the returned collection uses the same name as the variable, i.e. "result"? That would be easy to remember.

That seems reasonable.

**Pull request merged**